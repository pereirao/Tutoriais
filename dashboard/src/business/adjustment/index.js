// @flow

export type Adjustment = {
  id: number,
  created_at: string, // rails timestamp
  amount: number,
  credit: boolean,
  reason: string,
  description: string
};

export * as adjustment_actions from './actions';
export * as adjustment_epics from './epics';
export * as adjustment_helpers from './helpers';
export { default as adjustmentReducer } from './reducer';
export { default as adjustment_selectors } from './selectors';
