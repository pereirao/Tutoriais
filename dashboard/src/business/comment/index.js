// @flow
export type Author = {
  email: string
};

export type Comment = {
  id: number,
  created_at: string, // rails timestamp
  note: string,
  author: Author,
  posted_as: string
};

export * as comment_actions from './actions';
export * as comment_epics from './epics';
export * as comment_helpers from './helpers';
export { default as commentReducer } from './reducer';
export { default as comment_selectors } from './selectors';
