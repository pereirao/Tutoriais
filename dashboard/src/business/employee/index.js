// @flow

type EmployeeRole = 'supplier' | 'driver';

export type Employee = {
  id: number,
  first_name: string,
  last_name: string,
  active: boolean,
  email: string,
  roles: Array<EmployeeRole>
};

export * as employee_epics from './epics';
export * as employee_actions from './actions';
export * as employee_helpers from './helpers';
export { default as employeeReducer } from './reducer';
export { default as employee_selectors } from './selectors';
