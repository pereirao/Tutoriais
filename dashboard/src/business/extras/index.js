export * from './extras.dux';
export * from './extras.sagas';
