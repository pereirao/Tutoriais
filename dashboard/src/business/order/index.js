// @flow

export type OrderTags = {
  gift: boolean,
  out_of_hours: boolean,
  vip: boolean,
  corporate: boolean,
  scheduled: boolean,
  new_customer: boolean
};

export type BillingInfo = {
  name: string,
  card_type: string,
  card_last4: string
};

export type Address = {
  address1: string,
  address2: string,
  city: string,
  state: string,
  zip_code: string,
  coords: {lat: number, lng: number}
};

export type PickupDetail = {
  id: string,
  name: string,
  phone: string
};

export type Amounts = {
  total: number,
  subtotal: number,
  tax: number,
  tip: number,
  minibar_promos: number,
  store_discounts: number,
  discounts?: number, // FIXME: optional until sunset
  delivery_fee: number,
};

export type DeliveryService = {
  id: number,
  name: string,
  email: string,
  created_at: string,
  updated_at: string,
  minimal_delivery_time: number
}

export type DeliveryMethodType = 'on_demand' | 'shipped' | 'pickup';

export type DeliveryMethod = {
  type: DeliveryMethodType,
  expectation: string
};

export type AttributeFilters = 'vip' | 'gift' | 'scheduled' | 'exception';

export type OrderFilters = {
  date_range: {
    start: string, // iso-8601 timestamp
    end: string // iso-8601 timestamp
  },
  delivery_method_types: Array<DeliveryMethodType>,
  attributes: Array<AttributeFilters>
}

export type ExceptionMetadata = {
  order_item_sku?: string,
  replacement_item?: string,
  replacement_price?: string,
  correct_price?: string,
  incorrect_listing?: string,
  incomplete_address?: string,
  suspected_fraud?: string,
  failed_delivery?: string,
  driver_unavailable_why?: string,
  other?: string,
};

export type ExceptionType = 'out_of_stock' |
  'incorrect_pricing' |
  'incorrect_listing' |
  'incomplete_address' |
  'suspected_fraud' |
  'driver_unavailable' |
  'failed_delivery' |
  'other';

export type Exception = {
  type: ExceptionType,
  description: string,
  metadata: ExceptionMetadata
};

export const PROBLEM_TYPES = [
  'out_of_stock',
  'incorrect_pricing',
  'incorrect_listing',
  'incomplete_address',
  'suspected_fraud',
  'driver_unavailable',
  'failed_delivery',
  'other'
];

export type OrderListName = 'all' | 'active' | 'completed' | 'future';

export type RecipientInfo = { short_name: string, long_name: string, phone: string };

export type TimelineEntryType = 'created_at'
  | 'confirmed_at'
  | 'en_route_at'
  | 'delivered_at'
  | 'canceled_at'
  | 'comment';

export type TimelineEntry = {
  time: string,
  type: TimelineEntryType,
  meta?: { name?: string, driverName?: string }
}

export type OrderState = 'paid'
  | 'confirmed'
  | 'scheduled'
  | 'canceled'
  | 'exception'
  | 'en_route'
  | 'delivered';

export type OrderItem = {
  id: number,
  sku: string,
  name: string,
  item_size: string,
  unit_price: number,
  quantity: number
};

export type Order = {
  id: number,
  number: number,
  created_at: string, // rails timestamp
  confirmed_at: string, // rails timestamp
  canceled_at: string, // rails timestamp
  delivered_at: string, // rails timestamp
  scheduled_for: string, // rails timestamp
  scheduled_for_end: string, // rails timestamp
  order_time: string, // rails timestamp
  birthdate: string,
  notes: string,
  delivery_method: DeliveryMethod,
  state: OrderState,
  gift_message: string,
  recipient_info: RecipientInfo,
  customer_name: string,
  billing: BillingInfo,
  address?: Address,
  pickup_detail?: PickupDetail,
  amounts: Amounts,
  type_tags: OrderTags,
  receipt_url: string,
  order_items: Array<OrderItem>,
  comment_ids: Array<number>, // NOTE: shouldn't be hydrated
  order_adjustment_ids: Array<number>, // NOTE: shouldn't be hydrated
  driver_id: number,
  driver?: any,
  using_delivery_service: boolean,
  delivery_service: DeliveryService
};

export * as order_actions from './actions';
export * as order_helpers from './helpers';
export * as order_epics from './epics';
export { default as orderReducer } from './reducer';
export { default as order_selectors } from './selectors';
