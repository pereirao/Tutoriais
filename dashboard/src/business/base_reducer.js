//@flow

import {combineReducers} from 'redux';
import {adjustmentReducer} from './adjustment';
import {commentReducer} from './comment';
import {employeeReducer} from './employee';
import {notificationReducer} from './notification';
import {notificationMethodReducer} from './notification_method';
import {orderReducer} from './order';
import {reportReducer} from './report';
import {uiReducer} from './ui';
import {sessionReducer} from './session';
import {variantReducer} from './variant';
import {settingsReducer} from './settings';
import substituteReducer from './substitute/substitute.dux';
import productReducer from './product/product.dux';
import {connectRouter} from "connected-react-router";

const appReducer = (history) => combineReducers({
  adjustment: adjustmentReducer,
  comment: commentReducer,
  employee: employeeReducer,
  notification: notificationReducer,
  notification_method: notificationMethodReducer,
  order: orderReducer,
  report: reportReducer,
  ui: uiReducer,
  router: connectRouter(history),
  session: sessionReducer,
  variant: variantReducer,
  settings: settingsReducer,
  substitute: substituteReducer,
  product: productReducer
});

const rootReducer = (history) => (state, action) => {
  let app_state = state;
  if (action.type === 'SESSION:RESET_STATE') {
    app_state = undefined;
  }
  return appReducer(history)(app_state, action);
};

export default rootReducer;
