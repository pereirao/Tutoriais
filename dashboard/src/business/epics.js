import _ from 'lodash';

import { combineEpics } from 'redux-observable';

import { adjustment_epics } from './adjustment';
import { comment_epics } from './comment';
import { order_epics } from './order';
import { session_epics } from './session';
import { employee_epics } from './employee';
import { notification_method_epics } from './notification_method';
import { report_epics } from './report';
import { variant_epics } from './variant';
import { settings_epics } from './settings';

// convert the imported epic objects into an array
const epics_by_module = [
  adjustment_epics,
  comment_epics,
  order_epics,
  session_epics,
  employee_epics,
  notification_method_epics,
  report_epics,
  variant_epics,
  settings_epics
];

const epics = _.flatMap(epics_by_module, (module_epic_exports) => {
  // we pull the actual epic functions out of the exports we recieved from the each file
  const module_epics = _.values(module_epic_exports);
  return module_epics;
});

const rootEpic = combineEpics(...epics);
export default rootEpic;
