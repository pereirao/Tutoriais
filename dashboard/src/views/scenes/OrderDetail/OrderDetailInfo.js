// @flow

import _ from 'lodash';
import React from 'react';
import I18n from '../../../localization';
import type Order from '../../../business/order';
import {order_helpers} from '../../../business/order';
import style_sheet from './OrderDetail.module.css';
import OrderDetailInfoMapLinks from './OrderDetailInfoMapLinks';

const {
  cardDetails,
  customerName,
  customerNotes,
  recipientName,
  formattedAddress,
  addressLocation,
  customerBirthdate,
  recipientPhone,
  deliveryMethodType
} = order_helpers;

type OrderDetailInfoProps = { order: Order };
const OrderDetailInfo = ({order}: OrderDetailInfoProps) => (
  <div className={style_sheet.infoRow}>
    <div className={style_sheet.infoItem}>
      <div className={style_sheet.sectionHeader}><p
        className={style_sheet.sectionTitle}>{I18n.t(`ui.header.recipient.${deliveryMethodType(order)}`)}</p></div>
      <div className={style_sheet.infoItemContents}>
        <p className={style_sheet.recipientName}>{recipientName(order)}</p>
        <OrderAddress formatted_address={formattedAddress(order)} address_location={addressLocation(order)}/>
        <p>{recipientPhone(order)}</p>
      </div>
    </div>
    <div className={style_sheet.infoItem}>
      <div className={style_sheet.sectionHeader}><p
        className={style_sheet.sectionTitle}>{I18n.t('ui.header.ordered_by')}</p></div>
      <div className={style_sheet.infoItemContents}>
        <p>{customerName(order)}</p>
        <p>{cardDetails(order)}</p>
        {getBirthDate(order)}
      </div>
    </div>
    <CustomerNotesSection notes={customerNotes(order)}/>
  </div>
);
const getBirthDate = order => {
  if (order.birthdate) {
    return (
      <p>Birthdate: {customerBirthdate(order)}</p>
    );
  }
  return undefined;
};

type OrderAddressProps = { formatted_address?: { [main_text]: string, [secondary_text]: string, [latitude]: float, [longitude]: float }, address_location?: { [latitude]: float, [longitude]: float } };
export const OrderAddress = ({formatted_address, address_location}: OrderAddressProps) => {
  if (!formatted_address) return null;

  return (
    <div className={style_sheet.addressLineWrapper}>
      <p className={style_sheet.addressLine}>{formatted_address.main_text}</p>
      <p className={style_sheet.addressLine}>{formatted_address.secondary_text}</p>
      <OrderDetailInfoMapLinks coords={address_location}/>
    </div>
  );
};

type CustomerNotesSectionProps = { notes: string };
export const CustomerNotesSection = ({notes}: CustomerNotesSectionProps) => {
  if (_.isEmpty(notes)) return null;
  return (
    <div className={style_sheet.infoItem}>
      <div className={style_sheet.sectionHeader}><p
        className={style_sheet.sectionTitle}>{I18n.t('ui.header.customer_notes')}</p></div>
      <div className={style_sheet.infoItemContents}>
        <p>{notes}</p>
      </div>
    </div>
  );
};
export default OrderDetailInfo;
