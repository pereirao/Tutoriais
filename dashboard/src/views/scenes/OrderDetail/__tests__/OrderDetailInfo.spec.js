import React from 'react';

import order_factory from '../../../../business/order/__tests__/order.factory';
import { order_helpers } from '../../../../business/order';
import OrderDetailInfo, { OrderAddress, CustomerNotesSection } from '../OrderDetailInfo';

const { formattedAddress, customerNotes } = order_helpers;

it('renders OrderDetailInfo correctly', () => {
  expect(shallow(<OrderDetailInfo order={order_factory.build()} />)).toMatchSnapshot();
});

it('omits customer notes if not present', () => {
  expect(shallow(<OrderDetailInfo order={order_factory.build({notes: ''})} />)).toMatchSnapshot();
});

it('omits recipient address if pickup order', () => {
  expect(shallow(<OrderDetailInfo order={order_factory.build('pickup')} />)).toMatchSnapshot();
});

it('renders OrderAddress correctly', () => {
  expect(shallow(<OrderAddress formattedAddress={formattedAddress(order_factory.build())} />)).toMatchSnapshot();
});

it('renders CustomerNotesSection correctly', () => {
  expect(shallow(<CustomerNotesSection formattedAddress={customerNotes(order_factory.build())} />)).toMatchSnapshot();
});
