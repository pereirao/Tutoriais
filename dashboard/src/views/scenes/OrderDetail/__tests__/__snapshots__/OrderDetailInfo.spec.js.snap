// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`omits customer notes if not present 1`] = `
<div
  className="infoRow"
>
  <div
    className="infoItem"
  >
    <div
      className="sectionHeader"
    >
      <p
        className="sectionTitle"
      >
        DELIVER TO
      </p>
    </div>
    <div
      className="infoItemContents"
    >
      <p
        className="recipientName"
      >
        Biggie Smalls - East Coast
      </p>
      <OrderAddress
        formatted_address={
          Object {
            "main_text": "21 Jump Street, Apt 5",
            "secondary_text": "Brooklyn, NY 11206",
          }
        }
      />
      <p>
        (555) 555-5555
      </p>
    </div>
  </div>
  <div
    className="infoItem"
  >
    <div
      className="sectionHeader"
    >
      <p
        className="sectionTitle"
      >
        ORDERED BY
      </p>
    </div>
    <div
      className="infoItemContents"
    >
      <p>
        Biggie Smalls
      </p>
      <p>
        Amex **** 4321
      </p>
    </div>
  </div>
  <CustomerNotesSection
    notes=""
  />
</div>
`;

exports[`omits recipient address if pickup order 1`] = `
<div
  className="infoRow"
>
  <div
    className="infoItem"
  >
    <div
      className="sectionHeader"
    >
      <p
        className="sectionTitle"
      >
        PICKUP BY 
      </p>
    </div>
    <div
      className="infoItemContents"
    >
      <p
        className="recipientName"
      >
        Biggie Smalls - East Coast
      </p>
      <OrderAddress
        formatted_address={null}
      />
      <p>
        (555) 555-5555
      </p>
    </div>
  </div>
  <div
    className="infoItem"
  >
    <div
      className="sectionHeader"
    >
      <p
        className="sectionTitle"
      >
        ORDERED BY
      </p>
    </div>
    <div
      className="infoItemContents"
    >
      <p>
        Biggie Smalls
      </p>
      <p>
        Amex **** 4321
      </p>
    </div>
  </div>
  <CustomerNotesSection
    notes="Ring the doorbell"
  />
</div>
`;

exports[`renders CustomerNotesSection correctly 1`] = `null`;

exports[`renders OrderAddress correctly 1`] = `null`;

exports[`renders OrderDetailInfo correctly 1`] = `
<div
  className="infoRow"
>
  <div
    className="infoItem"
  >
    <div
      className="sectionHeader"
    >
      <p
        className="sectionTitle"
      >
        DELIVER TO
      </p>
    </div>
    <div
      className="infoItemContents"
    >
      <p
        className="recipientName"
      >
        Biggie Smalls - East Coast
      </p>
      <OrderAddress
        formatted_address={
          Object {
            "main_text": "21 Jump Street, Apt 5",
            "secondary_text": "Brooklyn, NY 11206",
          }
        }
      />
      <p>
        (555) 555-5555
      </p>
    </div>
  </div>
  <div
    className="infoItem"
  >
    <div
      className="sectionHeader"
    >
      <p
        className="sectionTitle"
      >
        ORDERED BY
      </p>
    </div>
    <div
      className="infoItemContents"
    >
      <p>
        Biggie Smalls
      </p>
      <p>
        Amex **** 4321
      </p>
    </div>
  </div>
  <CustomerNotesSection
    notes="Ring the doorbell"
  />
</div>
`;
